proposals/0723-qualified-strings.rst

```diff
@@ -0,0 +1,289 @@
+Qualified Strings
+=================
+
+.. author:: Brandon Chinn
+.. date-accepted:: Leave blank. This will be filled in when the proposal is accepted.
+.. ticket-url:: Leave blank. This will eventually be filled with the
+                ticket URL which will track the progress of the
+                implementation of the feature.
+.. implemented:: Leave blank. This will be filled in with the first GHC version which
+                 implements the described feature.
+.. highlight:: haskell
+.. header:: This proposal is `discussed at this pull request <https://github.com/ghc-proposals/ghc-proposals/pull/723>`_.
+.. sectnum::
+.. contents::
+
+This proposal proposes extending ``-XQualifiedDo`` to literal strings, to enable more ergonomic and more powerful syntax than ``OverloadedStrings``. Another way to view this proposal would be extending ``-XRebindableSyntax`` to literal strings, but only within a local scope.
+
+See also:
+
+* `QualifiedLists <https://github.com/ghc-proposals/ghc-proposals/pull/724>`_
+* `QualifiedNumerics <https://github.com/ghc-proposals/ghc-proposals/pull/725>`_
+
+Motivation
+----------
+
+Problems with Type Class-driven overloading
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+``OverloadedStrings`` works by desugaring to a type-class-overloaded function. That leads to certain general shortcomings:
+
+* It is a module-wide setting
+
+  * Anecdotally, people would rather avoid ``OverloadedStrings`` than deal with overloaded strings in the entire module.
+
+  * It's possible that this is one reason this extension isn't a default in GHC202X language editions, despite being in GHC for a long time.
+
+* Type inference ambiguity.
+
+  * Consider the following code:
+
+    ::
+
+      output :: IsString s => s -> IO ()
+
+      main = do
+        output "hello"
+        output "world"
+
+        -- output $ T.replace " " "_" input
+
+    This originally works with no extensions, due to the string literals being typed to concrete ``String``. But say the developer wants to add a call to ``T.replace`` and use ``Text`` literals; adding ``OverloadedStrings`` would cause ambiguity to the existing locations because they are now no longer concretely ``String``.
```

Comment:
I found it very hard to understand this example because you don't give a type signature for `replace`, nor `input`.  

The commented-out bit is also confusing.  Better to say
```
    output T.replace " " "_" input    -- Rejected by type checker (ambiguity)

Reply:
How's that?